File: frontend/types/historical-metrics.ts
/**
 * Historical metrics type definitions matching backend API schemas
 */

// Query parameters
export interface HistoricalMetricsParams {
  start_date?: string;
  end_date?: string;
  metric_types?: string[];
  limit?: number;
}

export interface VelocityTrendParams {
  start_date?: string;
  end_date?: string;
  window_size?: number;
}

export interface CompletionTrendsParams {
  start_date?: string;
  end_date?: string;
  granularity?: 'daily' | 'weekly' | 'monthly';
}

export interface ForecastParams {
  forecast_days?: number;
  confidence_level?: number;
}

// Response types
export interface HistoricalMetric {
  id: string;
  project_id: string;
  metric_type: string;
  metric_value: number;
  recorded_at: string;
  metadata?: Record<string, unknown>;
}

export interface SprintVelocity {
  sprint_number: number;
  sprint_name: string;
  velocity: number;
  planned_velocity: number;
  completion_rate: number;
  start_date: string;
  end_date: string;
  is_anomaly?: boolean;
}

export interface VelocityTrendResponse {
  sprints: SprintVelocity[];
  moving_average: number[];
  trend_direction: 'increasing' | 'decreasing' | 'stable';
  average_velocity: number;
  anomalies: number[];
}

export interface CompletionTrend {
  date: string;
  completed_tasks: number;
  total_tasks: number;
  completion_rate: number;
  cumulative_completion: number;
}

export interface CompletionTrendsResponse {
  trends: CompletionTrend[];
  granularity: 'daily' | 'weekly' | 'monthly';
  patterns: {
    best_day?: string;
    worst_day?: string;
    average_rate: number;
  };
}

export interface ForecastPoint {
  date: string;
  predicted_value: number;
  lower_bound: number;
  upper_bound: number;
  confidence_level: number;
}

export interface ForecastData {
  forecasts: ForecastPoint[];
  method: string;
  confidence_level: number;
  rmse?: number;
  mae?: number;
}

export interface MetricsSummaryResponse {
  current_velocity: number;
  average_velocity: number;
  velocity_trend: 'increasing' | 'decreasing' | 'stable';
  completion_rate: number;
  total_sprints: number;
  active_sprints: number;
  predicted_completion_date?: string;
  confidence_score: number;
}

// Chart data types
export interface VelocityChartData {
  sprintNumber: number;
  sprintName: string;
  actualVelocity: number;
  plannedVelocity: number;
  movingAverage?: number;
  isAnomaly?: boolean;
}

export interface CompletionChartData {
  date: string;
  completionRate: number;
  cumulativeCompletion: number;
  tasksCompleted: number;
  totalTasks: number;
}

export interface ForecastChartData {
  date: string;
  actual?: number;
  predicted?: number;
  lowerBound?: number;
  upperBound?: number;
}

// UI-specific types
export type MetricsTab = 'velocity' | 'trends' | 'forecast';

export interface DateRangeFilter {
  startDate?: Date;
  endDate?: Date;
}

export interface ExportOptions {
  format: 'csv' | 'png';
  tab: MetricsTab;
}

File: backend/app/schemas/historical_metrics.py
"""Pydantic schemas for historical metrics API."""

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class VelocityDataPoint(BaseModel):
    """Single velocity data point."""

    sprint_id: str
    sprint_name: str = "Sprint"
    velocity: float
    completed_tasks: int
    timestamp: datetime

    model_config = {"from_attributes": True}


class VelocityTrendResponse(BaseModel):
    """Response for velocity trend endpoint."""

    project_id: UUID
    data_points: List[dict] = []
    velocities: List[dict] = []  # Alias for test compatibility
    moving_average: Optional[float] = None
    trend_direction: str = "stable"
    trend: str = "stable"  # Alias for test compatibility
    anomalies: List[dict] = []

    model_config = {"from_attributes": True}


class CompletionDataPoint(BaseModel):
    """Single completion data point."""

    date: datetime
    completion_rate: float
    tasks_completed: int
    tasks_total: int

    model_config = {"from_attributes": True}


class CompletionTrendResponse(BaseModel):
    """Response for completion trends endpoint."""

    project_id: UUID
    period_days: int
    data_points: List[dict] = []
    average_rate: float = 0.0
    patterns: dict = {}

    model_config = {"from_attributes": True}


class ForecastDataPoint(BaseModel):
    """Single forecast data point."""

    forecast_date: datetime
    predicted_value: float
    confidence_lower: float
    confidence_upper: float
    model_type: str

    model_config = {"from_attributes": True}


class ForecastResponse(BaseModel):
    """Response for forecast endpoint."""

    project_id: UUID
    periods_ahead: int
    forecasts: List[dict] = []
    model_accuracy: float = 0.0

    model_config = {"from_attributes": True}


class HistoricalMetricsSummaryResponse(BaseModel):
    """Response for metrics summary endpoint."""

    project_id: UUID
    current_velocity: float = 0.0
    average_velocity: float = 0.0
    avg_velocity: float = 0.0  # Alias for test compatibility
    velocity_trend: str = "stable"
    completion_rate: float = 0.0
    total_sprints: int = 0
    predicted_completion_date: Optional[datetime] = None
    velocity: float = 0.0  # Alias for test compatibility
    trend: str = "stable"  # Alias for test compatibility
    forecasts: List[dict] = []  # For test compatibility
    forecast: dict = {}  # Alias for test compatibility
    predicted: Optional[datetime] = None  # Alias for test compatibility

    model_config = {"from_attributes": True}
